docs/validation.md

```diff
@@ -616,21 +616,220 @@ While this behaviour is unexpected (issues [#129](https://github.com/ajv-validat
 
 ### Removing unevaluated properties
 
-With [option `removeUnevaluated`](./api.md#options) you can filter data during the validation.
+With [option `removeUnevaluated`](./api.md#options) you can filter data during the validation. It is only compatible with draft 2019-09.
+
+This option modifies original data. It removes every property or item that would raise an error based on the keywords `unevaluatedProperties` and `unevaluatedItems`. It will first remove unevaluated properties then unevaluated items.
+
+Compared to `removeAdditional`, it will always take into account all sub-schemas that were valid for this data instance : 
+
+
+- all subschemas schemas in `allOf` and `$ref` keywords
+- valid sub-schemas in `oneOf` and `anyOf` keywords
+- sub-schema in `if` keyword
+- sub-schemas in `then` or `else` keywords that were applied based on the validation result by `if` keyword.
+
+
+
+
+Example 1 : remove unevaluated properties 
+
+```javascript
+
+const ajv = new Ajv({removeUnevaluated: true})
+
+const schema = {
+  $schema: "http://json-schema.org/draft/2019-09/schema#",
+  definitions: {
+    reference1: {
+      type: "object",
+      properties: { a: { type: "string" } }
+    },
+    reference2: {
+      type: "object",
+      properties: { a: { type: "string" } },
+      unevaluatedProperties: true
+    }
+  },
+  unevaluatedProperties: false,
+  type: "object",
+  properties: {
+    foo: { type: "string" }
+  },
+  required: ["foo"],
+  if: {
+    properties: { foo: { const: "A" } }
+  },
+  then: {
+    $ref: "#/definitions/reference1"
+  },
+  else: {
+    $ref: "#/definitions/reference2"
+  }
+};
+
+const validate = ajv.compile(schema)
+
+const data1 = {
+  foo: "A",
+  a: "", 
+  b: 1 // will be removed; `unevaluatedProperties` == false
+}
+
+console.log(validate(data1)) // true
+console.log(data1); // { foo: "A", a: "" }
+
+
+const data2 = {
+  foo: "",
+  a: "", 
+  b: 1 // will NOT be removed; `unevaluatedProperties` != false
+}
+
+console.log(validate(data2)) // true
+console.log(data2); // { foo: "A", a: "", b: 1 }
+
+```
+
+Example 2 : remove unevaluated properties with combination keyword
+
+```javascript
+
+const ajv = new Ajv({removeUnevaluated: true})
+
+const schema = {
+  $schema: "http://json-schema.org/draft/2019-09/schema#",
+  type: "object",
+  properties: {
+    foo: {
+      type: "object",
+      oneOf: [
+        {
+          properties: {
+            x: { type: "string" }
+          },
+          required: ["x"],
+          unevaluatedProperties: false
+        },
+        {
+          properties: {
+            y: { type: "integer" }
+          },
+          required: ["y"],
+          unevaluatedProperties: true
+        }
+      ]
+    }
+  },
+  required: ["foo"]
+};
+
+const validate = ajv.compile(schema)
+
+const data1 = {
```

Comment:
I am actually not sure data1 will pass the oneOf validation. oneOf means it needs to pass exclusively one of the validations. It might technically be considered failing on the first (bc of unevaluated) and passing on the second. But barring the unevaluated, it would be passing on both, which means it should fail. 

You have some great examples, I will update the tests with your examples

Reply:
I have update example 2.  {
    x: "", 
    y: 1 
  }

would actually pass since it matches the second oneOf validation.


{
    x: "", 
    z: 1 // will be removed; `unevaluatedProperties` == false
  }

However would not pass. It both raises an unevaluatedProperties error and a validation fail for the oneOf keyword. 

```javascript
[
  {
    "keyword": "unevaluatedProperties",
    "dataPath": "/foo",
    "schemaPath": "#/properties/foo/oneOf/0/unevaluatedProperties",
    "params": {
      "unevaluatedProperty": "z"
    },
    "message": "should NOT have unevaluated properties"
  },
  {
    "keyword": "required",
    "dataPath": "/foo",
    "schemaPath": "#/properties/foo/oneOf/1/required",
    "params": {
      "missingProperty": "y"
    },
    "message": "should have required property 'y'"
  },
  {
    "keyword": "oneOf",
    "dataPath": "/foo",
    "schemaPath": "#/properties/foo/oneOf",
    "params": {
      "passingSchemas": null
    },
    "message": "should match exactly one schema in oneOf"
  }
]
```

It is a pretty tricky example actually. 

- Either the object should fail the oneOf validation first and foremost, and we don't modify anything. Then I am not sure why ajv raises an error for every validation inside oneOf. This is probably the easiest and less error prone route.

- Or the object first raises an unevaluated property error, we remove this property, then the object becomes valid for the first oneOf validation. The problem is, why would we prefer "casting" the object to the first validation and ignore the second, since the object initially failed for both. Also I suspect the implementation would be harder to write.

Reply:
Depending on our understanding, this will also affect the outcome of Example 4.